nav2_simple_commander/nav2_simple_commander/footprint_collision_checker.py

```diff
@@ -0,0 +1,96 @@
+#! /usr/bin/env python3
+# Copyright 2021 Samsung Research America
+# Copyright 2022 Afif Swaidan
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from math import cos, sin
+from line_iterator import LineIterator
+from costmap_2d import PyCostmap2D
+from geometry_msgs.msg import Point
+
+NO_INFORMATION = 255
+LETHAL_OBSTACLE = 254
+INSCRIBED_INFLATED_OBSTACLE = 253
+MAX_NON_OBSTACLE = 252
+FREE_SPACE = 0
+
+
+class FootprintCollisionChecker():
+    def __init__(self, costmap: PyCostmap2D):
+        self.costmap_ = costmap
+
+    def footprintCost(self, footprint: list[Point]):
+        footprint_cost = 0.0
+        if not self.costmap_.worldToMapCheck(footprint[0].x, footprint[0].y):
+            return LETHAL_OBSTACLE
+
+        x0, y0 = self.costmap_.worldToMap(footprint[0].x, footprint[0].y)
+        xstart = x0
+        ystart = y0
+
+        for i in range(len(footprint) - 1):
+            if not self.costmap_.worldToMapCheck(
+                    footprint[i + 1].x, footprint[i + 1].y):
+                return LETHAL_OBSTACLE
+
+            x1, y1 = self.worldToMap(
+                footprint[i + 1].x, footprint[i + 1].y)
+
+            footprint_cost = max(self.lineCost(x0, x1, y0, y1), footprint_cost)
+            x0 = x1
+            y0 = y1
+
+            if footprint_cost == LETHAL_OBSTACLE:
+                return footprint_cost
+
+        return max(self.lineCost(xstart, x1, ystart, y1), footprint_cost)
+
+    def lineCost(self, x0, x1, y0, y1):
+        line_cost = 0.0
+        point_cost = -1.0
+        line_iterator = LineIterator(x0, y0, x1, y1, 0.1)
+
+        while line_iterator.isValid():
+            point_cost = self.pointCost(
+                line_iterator.getX(), line_iterator.getY())
+
+            if point_cost == LETHAL_OBSTACLE:
+                return point_cost
+
+            if line_cost < point_cost:
+                line_cost = point_cost
+
+        return line_cost
+
+    def worldToMap(self, wx: float, wy: float, mx: int, my: int):
+        return self.costmap_.worldToMap(wx, wy, mx, my)
+
+    def pointCost(self, x: int, y: int):
+        return self.costmap_.getCostXY(x, y)
+
+    def setCostmap(self, costmap: PyCostmap2D):
+        self.costmap_ = costmap
+
+    def footprintCostAtPose(self, x: float, y: float, theta: float, footprint: list[Point]):
```

Comment:
As well with Polygon message

Reply:
Also, a doc string would be nice here to explain the units (do we assume the X,Y,Theta are in world or map coords? Is the footprint in world or map coords?

Reply:
Alright noted,  I wanted to add the docstrings and the unit tests after I take your initial opinion.
I will replace the point message with the polygon message, thanks for the notice!